compiler/ballerina-lang/src/main/java/io/ballerina/projects/CompletionManager.java

```diff
@@ -225,4 +220,70 @@ public CompilerPluginInfo compilerPluginInfo() {
         }
     }
 
+    /**
+     * Visitor to validate the completion context.
+     *
+     */
+    static class ServiceDeclarationContextValidator extends NodeVisitor {
+
+        private CompletionContext context;
+        private boolean isValidContext = false;
+
+        public ServiceDeclarationContextValidator(CompletionContext context) {
+            this.context = context;
+        }
+
+        @Override
+        public void visit(ServiceDeclarationNode serviceDeclarationNode) {
+            isValidContext = true;
+        }
+
+        @Override
+        public void visit(SimpleNameReferenceNode simpleNameReferenceNode) {
+            simpleNameReferenceNode.parent().accept(this);
+        }
+
+        @Override
+        public void visit(ObjectFieldNode objectFieldNode) {
+            int cursorPosition = context.cursorPosInTree();
+            isValidContext = objectFieldNode.textRange().startOffset() <= cursorPosition
+                    && cursorPosition <= objectFieldNode.textRange().endOffset()
+                    && objectFieldNode.fieldName().isMissing()
+                    && objectFieldNode.equalsToken().isEmpty();
+        }
+
+        @Override
+        public void visit(FunctionDefinitionNode functionDefinitionNode) {
+            int cursorLine = context.cursorPosition().get().line();
+            isValidContext = cursorLine < functionDefinitionNode.lineRange().startLine().line()
+                    || functionDefinitionNode.lineRange().endLine().line() < cursorLine;
+        }
+
+        @Override
+        public void visit(MethodDeclarationNode methodDeclarationNode) {
+            int cursorLine = context.cursorPosition().get().line();
+            isValidContext = cursorLine < methodDeclarationNode.lineRange().startLine().line()
+                    || methodDeclarationNode.lineRange().endLine().line() < cursorLine;
+        }
+
+        @Override
+        protected void visitSyntaxNode(Node node) {
+            //Do nothing
+        }
+
+        public Boolean isValidContext() {
+            return this.isValidContext;
+        }
+
+        public void visitNode(Node node) {
+            if (node.kind() == SyntaxKind.LIST) {
+                if (node.parent() != null) {
+                    node.parent().accept(this);
+                }
+            } else {
+                node.accept(this);
+            }
+        }
```

Comment:
Instead of having this check, can't we implement the method for list constructor and visit the parent?

Reply:
This syntaxKind is for ExternalTreeNodeList, not for List contructor.